### PR TITLE
Native SpMV Kernel Update and nnz changes

### PR DESCRIPTION
**Pull request by mattkab2**
This merge updates the native SpMV kernel to require less communication by creating a communication map at assemble-time, meaning each processor knows the destination of incoming data as it appears when computing the matrix-vector product.

This merge also adds compatibility to the native types with the PETSc onnz/dnnz parameters which sidesteps a int/long conversion problem for the nnz parameterlist entry, which has a different meaning in the PETSc/Native case.

### CODE REVIEW

**Comment by mattkab2** [comment]
This merge now contains a patch to the CI which resolves a python error mismatch, preventing installation of pytables.
The CI is still running, but the line which caused the error (the pip install script) is no longer exiting with an error code.

The heavy tests in MPACT are being run with this merge, as well as serial-debug to ensure these changes don't cause errors. Results will be posted here when available.

**Comment by bkochuna** [comment]
Removing Shane from review. He's changed jobs. Adding Aaron instead.

File: src/MatrixTypes.f90
Comment (bkochuna) [inline src/MatrixTypes.f90:875]: Need to update comment block for change in parameters

File: src/MatrixTypes.f90
Comment (bkochuna) [inline src/MatrixTypes.f90:897]: I'm having a hard time understanding why this part of the IF-block `ANY(req == MPI_REQUEST_NULL)=TRUE` is needed.

File: src/MatrixTypes.f90
Comment (bkochuna) [inline src/MatrixTypes.f90:928]: Update this routine's header comment as well.

File: src/MatrixTypes_Native.f90
Comment (bkochuna) [inline src/MatrixTypes_Native.f90:508]: `kind` to `KIND`

**Review by bkochuna — COMMENTED** [review]
Just a light review.

File: src/MatrixTypes.f90
Comment (mattkab2) [reply src/MatrixTypes.f90:905]: The ANY(req == ...) is what controls whether the code searches for the empty slot (to return to the matrix multiplication code) or waits on ongoing communication to finish.
If any are MPI_REQUEST_NULL then that is an empty buffer and communication will start to that buffer.

I know we discussed this in our call but I'm putting it here if in the future, anyone has the same question.

**Review by aarograh — COMMENTED** [review]
I don't have anything to add to Brendan's review since he's more familiar with what's going on here than I am.  Once he approves I'm fine merging it.

**Comment by aarograh** [comment]
@mattkab2 looks like you managed to fix that stupid `pip` issue that's been plaguing our Futility CI for awhile.  Thanks for getting that one figured out!  I'm actually going to cherry-pick that change into a separate PR so we can get the CI cleaned up while you work on resolving Brendan's comments.

**Comment by mattkab2** [comment]
I've just realized that I never posted the update on the tests;

MPACT built with this branch passes all the heavies/serial debug tests that are currently passing on master.
There are a handful (~10) that are showing as failing for me on both, but those are mostly due to missing XS libraries (though there are a few failing COBRA_TF tests on master I have no explanation for).

**Comment by mattkab2** [comment]
@bkochuna @aarograh the MPACT build with this passes all the checks I can throw at it.
Could one of you review and merge when you get the chance?
Thanks.

**Comment by bkochuna** [comment]
This looks fine to me. I'm good with merging. The TRAVIS-CI did actually finish successfully (you can see it in details). Not sure why it hasn't updated here.

@aarograh anything else you want to see here?

**Comment by aarograh** [comment]
@bkochuna nope.  You can approve and merge if you deem it ready

File: src/MatrixTypes.f90
Comment (mattkab2) [reply src/MatrixTypes.f90:935]: This has been fixed.
For some reason, git didn't mark this as modified when I pushed up the last commit

File: src/MatrixTypes.f90
Comment (mattkab2) [reply src/MatrixTypes.f90:876]: This was fixed in the last commit. However, there is a type so I will fix this and push up, which will restart the pipeline and hopefully fix the 'stuck' pipeline status.

File: src/MatrixTypes_Native.f90
Comment (mattkab2) [reply src/MatrixTypes_Native.f90]: Ditto on this one, it's fixed in the newest version

**Review by bkochuna — APPROVED** [review]

**Pull request merged**
